### PR TITLE
🏗 New Travis Build Matrix

### LINKED ISSUE #20589: Implement build / test stages for `ampproject/amphtml` Travis builds
`ampproject/docs` already does this. We should do it too.

Documentation: https://docs.travis-ci.com/user/build-stages/
Sample build: https://travis-ci.org/ampproject/docs/builds/485520175
Config: https://github.com/ampproject/docs/blob/master/.travis.yml#L20-L38

More Travis VMs coming up in the next week or so.

### PR DESCRIPTION
**Pull request by estherkim**
This PR reorganizes the current build matrix to look like this:
image

Fixes #20589

### CODE REVIEW

File: build-system/pr-check/utils.js [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+/**
+ * Stops the timer for the given function and prints the execution time.
+ * @param {string} functionName
+ * @param {DOMHighResTimeStamp} startTime
+ * @return {number}
+ */
+function stopTimer(functionName, startTime) {
+  const endTime = Date.now();
+  const executionTime = new Date(endTime - startTime);
+  const mins = executionTime.getMinutes();
```
[the remark this answers is not in the document]
Comment (estherkim) [reply build-system/pr-check/utils.js]: True, this is copy pasta from build-system/pr-check.js. Can we punt?

File: build-system/pr-check/utils.js [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+const BUILD_OUTPUT_DIRS = 'build/ dist/ dist.3p/';
+const BUILD_OUTPUT_STORAGE_LOCATION = 'gs://amp-travis-builds';
+
+/**
+ * Starts connection to Sauce Labs after getting account credentials
+ * @param {string} functionName
+ */
+function startSauceConnect(functionName) {
+  process.env['SAUCE_USERNAME'] = 'amphtml';
+  process.env['SAUCE_ACCESS_KEY'] = getStdout('curl --silent ' +
```
[the remark this answers is not in the document]
Comment (estherkim) [reply build-system/pr-check/utils.js]: Copy pasta from build-system/pr-check.js. Punt to future PR?

File: build-system/pr-check/utils.js [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+const colors = require('ansi-colors');
+const {execOrDie, exec, getStdout} = require('../exec');
+const {travisBuildNumber} = require('../travis');
+const BUILD_OUTPUT_FILE = `amp_build_${travisBuildNumber()}.zip`;
```
[the remark this answers is not in the document]
Comment (estherkim) [reply build-system/pr-check/utils.js]: I haven't changed the way pr-check is run locally. No changes to build-system/pr-check.js so this should be okay. We'll need a future PR to clean up build-system/pr-check.js for sure.

File: build-system/pr-check/dist-test.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    timedExecOrDie('gulp bundle-size --on-push-build');
+    runSinglePassTest_();
+  }
+  else {
+    let runTests = buildTargets.has('RUNTIME');
+
+    unzipBuildOutput(); //TODO(estherkim): does this belong here?
+    timedExecOrDie('gulp dist --fortesting --noextensions');
+    timedExecOrDie('gulp bundle-size ' +
+    `${buildTargets.has('RUNTIME') ? '--on_pr_build' : '--on_skipped_build'}`);
```
[the remark this answers is not in the document]
Comment (estherkim) [reply build-system/pr-check/dist-test.js]: Cleaning up :)

File: build-system/pr-check/utils.js [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+      'echo travis_fold:start:verify_unzip_results && echo ${log} && ' +
+      'echo travis_fold:end:verify_unzip_results');
+}
+
+function zipBuildOutput() {
+  timedExecOrDie(
+      `log="$(zip -r ${BUILD_OUTPUT_FILE} ${BUILD_OUTPUT_DIRS})" && ` +
+      'echo travis_fold:start:zip_results && echo ${log} && ' +
+      'echo travis_fold:end:zip_results');
+  timedExecOrDie(`gsutil -m cp -r ${BUILD_OUTPUT_FILE} `
```
[the remark this answers is not in the document]
Comment (estherkim) [reply build-system/pr-check/utils.js]: Good point - added a check to see if it exists, if so, deletes.

File: build-system/pr-check/utils.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (danielrozenberg) [reply build-system/pr-check/utils.js]: :+1:

File: build-system/pr-check/dist-test.js [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  const startTime = startTimer(FILENAME);
+  const buildTargets = determineBuildTargets();
+
+  if (isTravisPushBuild()) {
+    timedExecOrDie('gulp dist --fortesting --noextensions');
+    timedExecOrDie('gulp bundle-size --on-push-build');
+    runSinglePassTest_();
+  }
+  else {
+    let runTests = false;
```
Comment (danielrozenberg) [inline build-system/pr-check/dist-test.js]: nit: `ranTests` - because you use it as an indicator to determine whether to show a message, not as a way to decide whether to run the tests

File: build-system/pr-check/utils.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (danielrozenberg) [reply build-system/pr-check/utils.js]: :+1:

File: build-system/pr-check/utils.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (danielrozenberg) [reply build-system/pr-check/utils.js]: :+1:

File: build-system/pr-check/utils.js [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+/**
+ * Starts a timer to measure the execution time of the given function.
+ * @param {string} functionName
+ * @return {DOMHighResTimeStamp}
+ */
+function startTimer(functionName) {
+  const startTime = Date.now();
+  const fileLogPrefix = colors.bold(colors.yellow(`${functionName}:`));
```
Comment (rsimha) [inline build-system/pr-check/utils.js]: `fileLogPrefix` is currently the name of the file in which the execution. It used to be `pr-check.js`, but now it's the full gulp command, which looks goofy. See original implementation below:

https://github.com/ampproject/amphtml/blob/b84350059273cec5d8e683b3bfae073cf356dcd1/build-system/pr-check.js#L58-L63

For a quick fix, maybe take `fileLogPrefix` as an arg, and pass it in via a helper function that dynamically determines the name of the given file? (The prefix could be whatever you want it to be, it's merely meant to make Travis logs more readable.)

File: build-system/pr-check/utils.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (rsimha) [reply build-system/pr-check/utils.js]: I don't think this is necessary. I just tested, and can verify that re-uploading an existing will overwrite the previous upload. So restarting a given Travis build will create a new build output file.

File: build-system/pr-check/validator.js [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+ * @fileoverview
+ * This script runs validator tests.
+ * This is run during the CI stage = build; job = validator.
+ */
+
+const {determineBuildTargets} = require('./build-target');
+const {isTravisPushBuild} = require('../travis');
+const {startTimer, stopTimer, timedExecOrDie} = require('./utils');
+
+const FILENAME = 'validator.js';
```
Comment (rsimha) [inline build-system/pr-check/validator.js]: This job will be a no-op most of the time, and yet, there's the overhead of spinning up a VM to do nothing. 

One way to avoid this is to run a standalone `nodejs` script in `.travis.yml` to determine the build targets, and make the various jobs start up conditionally.

Definitely out of scope for this PR, but let's consider this as a future optimization, since it will save a whole lot of VM spin-ups in the long run.

File: build-system/pr-check/visual-diff-test.js [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    buildTargets.has('FLAG_CONFIG')) {
+      unzipBuildOutput();
+      process.env['PERCY_TOKEN'] = atob(process.env.PERCY_TOKEN_ENCODED);
+      timedExecOrDie('gulp visual-diff --nobuild');
+    }
+    else {
+      timedExecOrDie('gulp visual-diff --nobuild --empty');
+      console.log('Skipping visual diff tests because this commit does ' +
+      'not affect the runtime, build system, integration test files, ' +
+      'visual diff test files, or flag config files.');
```
Comment (rsimha) [inline build-system/pr-check/visual-diff-test.js]: Copy-pasta messed up the indentation here.

File: build-system/pr-check/validator.js (hunk 8 above)
Comment (estherkim) [reply build-system/pr-check/validator.js]: Agreed! Would like to utilize Travis's build conditions feature, so a job doesn't show as "succeeded" when it really should be "skipped" or hidden.

File: build-system/pr-check/utils.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (estherkim) [reply build-system/pr-check/utils.js]: Ah. When I tested it locally, it updates the contents of the zip file, but doesn't fully replace. Anyways I added an `rm` for in case.

[the remark this answers is not in the document]
Comment (estherkim) [reply build-system/pr-check/utils.js]: Ah nevermind! I mixed up who was doing the zipping and who was doing the storing. You're right @rsimha, this shouldn't be a problem.

File: .travis.yml [hunk 10]
```diff
@@ -14,40 +13,99 @@ notifications:
...
+      script:
+        - node build-system/pr-check/build.js
+    - stage: build
+      name: "Checks" 
+      script:        
+        - node build-system/pr-check/checks.js
+    - stage: build
+      name: "Validator"        
+      before_script:
+        - pip install --user protobuf # required by validator
```
Comment (rsimha) [inline .travis.yml]: nit: Inline comment is no longer necessary.

File: .travis.yml [hunk 11]
```diff
@@ -14,40 +13,99 @@ notifications:
...
+    - stage: test
+      name: "Local Tests"
+      before_install: 
+        - export DISPLAY=:99.0
+        - sh -e /etc/init.d/xvfb start
+        # Explicitly add yarn and gsutil install because this section overrides before_install parent section
+        - curl -o- -L https://yarnpkg.com/install.sh | bash
+        - export PATH="$HOME/.yarn/bin:$HOME/.config/yarn/global/node_modules/.bin:$PATH"
+        - pip install urllib3[secure]
+        - pip install gsutil --user
```
Comment (rsimha) [inline .travis.yml]: Remove this `before_install` section too.

File: .travis.yml [hunk 12]
```diff
@@ -14,40 +14,103 @@ notifications:
...
+    - stage: test
+      name: "Dist, Bundle Size, Single Pass Tests"
+      before_install:
+        - export CHROME_BIN=google-chrome-stable
+        - export DISPLAY=:99.0
+        # Explicitly add yarn and gsutil install because this section overrides before_install parent section
+        - curl -o- -L https://yarnpkg.com/install.sh | bash
+        - export PATH="$HOME/.yarn/bin:$HOME/.config/yarn/global/node_modules/.bin:$PATH"
+        - pip install urllib3[secure]
+        - pip install gsutil --user
```
[the remark this answers is not in the document]
Comment (estherkim) [reply .travis.yml]: Don't we need to keep `yarn` and `gsutil` installs for all of the jobs?

[the remark this answers is not in the document]
Comment (rsimha) [reply .travis.yml]: What I meant was the global `before_install` at the top should be sufficient.

(If the only reason this is needed is the `DISPLAY` and `xvfb` stuff, let's just move it to the top and try to remove in a later clean up PR.)

File: .travis.yml [hunk 13]
```diff
@@ -14,40 +14,87 @@ notifications:
...
+      addons:
+        chrome: stable
+        hosts:
+          - ads.localhost
+          - iframe.localhost
+          # Requested by some tests because they need a valid font host,
+          # but should not resolve in tests.
+          - fonts.googleapis.com
```
[the remark this answers is not in the document]
Comment (rsimha) [reply .travis.yml]: If `hosts` is needed for one or more jobs, let's just move the entire `addons` section to the top to avoid repeating it everywhere. (It's inexpensive, and makes the rest of this file look cleaner.)

File: build-system/pr-check/utils.js [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+      `log="$(ls -la ${BUILD_OUTPUT_DIRS})" && ` +
+      'echo travis_fold:start:verify_unzip_results && echo ${log} && ' +
+      'echo travis_fold:end:verify_unzip_results');
+}
+
+function zipBuildOutput() {
+  timedExecOrDie(
+      `log="$(zip -r ${BUILD_OUTPUT_FILE} ${BUILD_OUTPUT_DIRS})" && ` +
+      'echo travis_fold:start:zip_results && echo ${log} && ' +
+      'echo travis_fold:end:zip_results');
```
Comment (rsimha) [inline build-system/pr-check/utils.js]: The output for this looks weird: https://travis-ci.org/ampproject/amphtml/jobs/493476480#L743.

Maybe for now, just skip timing the operation, and simply call `execOrDie` from `exec.js`.

File: build-system/pr-check/utils.js [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+      `${BUILD_OUTPUT_STORAGE_LOCATION}/${BUILD_OUTPUT_FILE} ` +
+      `${BUILD_OUTPUT_FILE}`);
+  timedExecOrDie(
+      `log="$(unzip -o ${BUILD_OUTPUT_FILE})" && ` +
+      'echo travis_fold:start:unzip_results && echo ${log} && ' +
+      'echo travis_fold:end:unzip_results');
+  timedExecOrDie(
+      `log="$(ls -la ${BUILD_OUTPUT_DIRS})" && ` +
+      'echo travis_fold:start:verify_unzip_results && echo ${log} && ' +
+      'echo travis_fold:end:verify_unzip_results');
```
Comment (rsimha) [inline build-system/pr-check/utils.js]: There's no need to time the zip operations here, since the output looks weird. Just call `execOrDie` from `exec.js` and apply the log folding above and below the unzip commands.

File: build-system/pr-check/utils.js [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+function zipBuildOutput() {
```
Comment (rsimha) [inline build-system/pr-check/utils.js]: Rename this to `uploadBuildOutput`, since that's the more important operation.

File: build-system/pr-check/utils.js [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+ * @param {string} fileName
+ */
+function timedExecOrDie(cmd, fileName = 'utils.js') {
+  const startTime = startTimer(cmd, fileName);
+  execOrDie(cmd);
+  stopTimer(fileName, startTime);
+}
+
+
+function unzipBuildOutput() {
```
Comment (rsimha) [inline build-system/pr-check/utils.js]: Rename this to `downloadBuildOutput`, since that's the more important operation.

File: build-system/pr-check/build.js [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+  startTimer,
+  stopTimer,
+  timedExecOrDie: timedExecOrDieBase,
+  zipBuildOutput} = require('./utils');
+const {determineBuildTargets} = require('./build-target');
+const {getStderr} = require('../exec');
+const FILENAME = 'build.js';
+const FILELOGPREFIX = colors.bold(colors.yellow(`${FILENAME}:`));
+const timedExecOrDie =
+  (cmd, unusedFileName) => timedExecOrDieBase(cmd, FILENAME);
```
Comment (rsimha) [inline build-system/pr-check/build.js]: Hah! Clever! :+1:

File: build-system/tasks/presubmit-checks.js [hunk 19]
```diff
@@ -104,6 +104,14 @@ const forbiddenTerms = {
...
       'build-system/amp4test.js',
       'build-system/check-package-manager.js',
+      'build-system/pr-check/build.js',
+      'build-system/pr-check/checks.js',
+      'build-system/pr-check/dist-test.js',
+      'build-system/pr-check/local-test.js',
+      'build-system/pr-check/remote-test.js',
+      'build-system/pr-check/utils.js',
+      'build-system/pr-check/validator.js',
+      'build-system/pr-check/visual-diff-test.js',
```
Comment (rsimha) [inline build-system/tasks/presubmit-checks.js]: Grammar nit: For consistency, pluralize `tests` in all these filenames, just like `checks` and `utils` are pluralized.
```
      'build-system/pr-check/dist-tests.js',
      'build-system/pr-check/local-tests.js',
      'build-system/pr-check/remote-tests.js',
      'build-system/pr-check/validator-tests.js',
      'build-system/pr-check/visual-diff-tests.js',
```

File: build-system/tasks/presubmit-checks.js [hunk 20]
```diff
@@ -104,6 +104,14 @@ const forbiddenTerms = {
       'build-system/app.js',
       'build-system/amp4test.js',
       'build-system/check-package-manager.js',
+      'build-system/pr-check/build.js',
+      'build-system/pr-check/checks.js',
+      'build-system/pr-check/dist-test.js',
+      'build-system/pr-check/local-test.js',
+      'build-system/pr-check/remote-test.js',
+      'build-system/pr-check/utils.js',
+      'build-system/pr-check/validator.js',
```
Comment (rsimha) [inline build-system/tasks/presubmit-checks.js]: For consistency, `validator-tests.js` sounds better than `validator.js`.

File: build-system/pr-check/dist-test.js [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+  unzipBuildOutput} = require('./utils');
+const {determineBuildTargets} = require('./build-target');
+const {isTravisPushBuild} = require('../travis');
+
+const FILENAME = 'dist-test.js';
+const timedExecOrDie =
+  (cmd, unusedFileName) => timedExecOrDieBase(cmd, FILENAME);
+
+function runSinglePassTest_() {
+  timedExecOrDie('rm -R dist');
```
Comment (rsimha) [inline build-system/pr-check/dist-test.js]: Prefer `gulp clean`, since it cleans up other directories too.

File: build-system/pr-check/dist-test.js [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+    timedExecOrDie('gulp dist --fortesting --noextensions');
+    timedExecOrDie('gulp bundle-size --on-push-build');
+    runSinglePassTest_();
+  }
+  else {
+    let ranTests = false;
+
+    if (buildTargets.has('RUNTIME')) {
+
+      unzipBuildOutput(); //TODO(estherkim): does this belong here?
```
Comment (rsimha) [inline build-system/pr-check/dist-test.js]: Remove this line, since we're calling `gulp dist` below.

File: build-system/pr-check/utils.js [hunk 23]
```diff
@@ -0,0 +1,174 @@
...
+/**
+ * Executes the provided command and times it. The program terminates in case of
+ * failure.
+ * @param {string} cmd
+ * @param {string} fileName
+ */
+function timedExecOrDie(cmd, fileName = 'utils.js') {
+  const startTime = startTimer(cmd, fileName);
+  execOrDie(cmd);
+  stopTimer(fileName, startTime);
```
Comment (rsimha) [inline build-system/pr-check/utils.js]: This should be `stopTimer(cmd, startTime)`

See https://travis-ci.org/ampproject/amphtml/jobs/493484525#L713-L717

File: build-system/pr-check/utils.js (hunk 15 above)
Comment (rsimha) [reply build-system/pr-check/utils.js]: Alternative implementation:

```
function downloadBuildOutput() {
  const buildOutputDownloadUrl = `${BUILD_OUTPUT_STORAGE_LOCATION}/${BUILD_OUTPUT_FILE}`;
  console.log(fileLogPrefix,
      'Downloading build output from',
      colors.cyan(buildOutputDownloadUrl) + '...');
  execOrDie(`gsutil cp ${buildOutputDownloadUrl} ${BUILD_OUTPUT_FILE}`);
  console.log(fileLogPrefix,
      'Extracting',
      colors.cyan(BUILD_OUTPUT_FILE) + '...');
  exec('echo travis_fold:start:unzip_results');
  execOrDie(`unzip -o ${BUILD_OUTPUT_FILE}`);
  exec('echo travis_fold:end:unzip_results');
  console.log(fileLogPrefix, 'Verifying extracted files...');
  exec('echo travis_fold:start:verify_unzip_results');
  execOrDie(`ls -la ${BUILD_OUTPUT_DIRS}`);
  exec('echo travis_fold:end:verify_unzip_results');
}
```

File: .travis.yml (hunk 12 above)
[the remark this answers is not in the document]
Comment (estherkim) [reply .travis.yml]: Got it got it

File: build-system/pr-check/utils.js (hunk 14 above)
Comment (estherkim) [reply build-system/pr-check/utils.js]: Done!

File: build-system/pr-check/utils.js (hunk 15 above)
Comment (estherkim) [reply build-system/pr-check/utils.js]: Done!

File: build-system/pr-check/utils.js [hunk 24]
```diff
@@ -0,0 +1,205 @@
...
+ */
+function downloadBuildOutput(functionName) {
+  const fileLogPrefix = colors.bold(colors.yellow(`${functionName}:`));
+  const buildOutputDownloadUrl =
+    `${BUILD_OUTPUT_STORAGE_LOCATION}/${BUILD_OUTPUT_FILE}`;
+
+  console.log(
+      `${fileLogPrefix} Downloading build output from ` +
+      `${colors.cyan(buildOutputDownloadUrl)}...`);
+  exec('echo travis_fold:start:download_results');
```
Comment (rsimha) [inline build-system/pr-check/utils.js]: Log readability nit: Add a newline after each instance of `travis_fold:start:foo` so that partial content doesn't show.

(See https://travis-ci.org/ampproject/amphtml/jobs/493824868#L751)

File: build-system/pr-check/utils.js [hunk 25]
```diff
@@ -0,0 +1,205 @@
...
+ * Zips and uploads the build output to a remote storage location
+ * @param {string} functionName
+ */
+function uploadBuildOutput(functionName) {
+  const fileLogPrefix = colors.bold(colors.yellow(`${functionName}:`));
+
+  console.log(
+      `\n${fileLogPrefix} Compressing contents of directories ` +
+      `${colors.cyan(BUILD_OUTPUT_DIRS)} into ` +
+      `${colors.cyan(BUILD_OUTPUT_FILE)}...`);
```
Comment (rsimha) [inline build-system/pr-check/utils.js]: Style nit: Don't include color formatting inside `${}`. Prefer `cyan(${foo})` to `${cyan(foo)}`, here and elsewhere in the PR.

Also, on reading the logs, there are files and directories being compressed, so let's rewrite this line as:
```
  const outputDirList = BUILD_OUTPUT_DIRS.split(' ').join(',');
  console.log(
      `\n${fileLogPrefix} Compressing ` +
      `colors.cyan(${outputDirList}) into ` +
      `colors.cyan(${BUILD_OUTPUT_FILE})...`);
```

(See https://travis-ci.org/ampproject/amphtml/jobs/493824868#L751)

File: build-system/pr-check/utils.js [hunk 26]
```diff
@@ -0,0 +1,205 @@
...
+function downloadBuildOutput(functionName) {
+  const fileLogPrefix = colors.bold(colors.yellow(`${functionName}:`));
+  const buildOutputDownloadUrl =
+    `${BUILD_OUTPUT_STORAGE_LOCATION}/${BUILD_OUTPUT_FILE}`;
+
+  console.log(
+      `${fileLogPrefix} Downloading build output from ` +
+      `${colors.cyan(buildOutputDownloadUrl)}...`);
+  exec('echo travis_fold:start:download_results');
+  execOrDie(`gsutil cp ${buildOutputDownloadUrl} ${BUILD_OUTPUT_FILE}`);
```
Comment (rsimha) [inline build-system/pr-check/utils.js]: TODO: Once this is merged, I suggest creating a dummy PR to test the `orDie` behavior by uploading to a non-existent storage location, and downloading a non-existent filename.

File: build-system/tasks/presubmit-checks.js [hunk 27]
```diff
@@ -397,6 +405,7 @@ const forbiddenTerms = {
   'isTrustedViewer': {
     message: requiresReviewPrivacy,
     whitelist: [
+      'extensions/amp-bind/0.1/bind-impl.js',
```
Comment (rsimha) [inline build-system/tasks/presubmit-checks.js]: Seems like this is an unrelated change. Revert.

File: .travis.yml [hunk 28]
```diff
@@ -33,21 +27,66 @@ branches:
...
+      addons:
+        chrome: stable
+      script:
+        - node build-system/pr-check/local-tests.js
+    - stage: test
+      name: "Remote (Sauce Labs) Tests"
+      script:
+        - node build-system/pr-check/remote-tests.js
+      after_script:
+        - build-system/sauce_connect/stop_sauce_connect.sh
```
Comment (rsimha) [inline .travis.yml]: This section looks great :)

File: .travis.yml [hunk 29]
```diff
@@ -33,21 +27,66 @@ branches:
...
+  - test
+jobs:
+  fast_finish: true
+  allowed_failures:
+    - stage: test
+      name: "Remote (Sauce Labs) Tests"
+      script:
+        - node build-system/pr-check/remote-tests.js
+      after_script:
+        - build-system/sauce_connect/stop_sauce_connect.sh
```
Comment (rsimha) [inline .travis.yml]: For easy maintainability, move this section to the bottom, so that when we make the Sauce Labs job blocking once again, all one has to do is remove the `allowed_failures` header.

File: build-system/pr-check/build.js [hunk 30]
```diff
@@ -0,0 +1,116 @@
...
+
+/**
+ * Makes sure package.json and yarn.lock are in sync.
+ * @private
+ */
+function runYarnIntegrityCheck_() {
+  const yarnIntegrityCheck = getStderr('yarn check --integrity').trim();
```
Comment (rsimha) [inline build-system/pr-check/build.js]: TODO: After this is merged, create a dummy PR that contains a bad yarn state (edit `package.json` but not `yarn.lock`) and make sure this check fails the build.

File: build-system/pr-check/checks.js [hunk 31]
```diff
@@ -0,0 +1,81 @@
...
+  printChangeSummary(FILENAME);
+
+  timedExecOrDie('gulp update-packages');
+  timedExecOrDie('gulp presubmit');
+  timedExecOrDie('gulp lint');
+
+  if (!isTravisPullRequestBuild()) {
+    timedExecOrDie('gulp ava');
+    timedExecOrDie('node node_modules/jest/bin/jest.js');
+    timedExecOrDie('gulp check-types');
```
Comment (rsimha) [inline build-system/pr-check/checks.js]: `check-types` has the longest running time, so move it to the bottom after `dep-check` so this job will fail faster on average.

File: build-system/pr-check/checks.js [hunk 32]
```diff
@@ -0,0 +1,81 @@
...
+      timedExecOrDie('gulp ava');
+      timedExecOrDie('node node_modules/jest/bin/jest.js');
+    }
+
+    if (buildTargets.has('RUNTIME') ||
+        buildTargets.has('BUILD_SYSTEM') ||
+        buildTargets.has('UNIT_TEST') ||
+        buildTargets.has('INTEGRATION_TEST')) {
+
+      timedExecOrDie('gulp check-types');
```
Comment (rsimha) [inline build-system/pr-check/checks.js]: Same here, move `check-types` to after `dep-check`.

File: build-system/pr-check/dist-tests.js [hunk 33]
```diff
@@ -0,0 +1,82 @@
...
+
+const FILENAME = 'dist-tests.js';
+const timedExecOrDie =
+  (cmd, unusedFileName) => timedExecOrDieBase(cmd, FILENAME);
+
+function runSinglePassTest_() {
+  timedExecOrDie('gulp clean');
+  timedExecOrDie('gulp dist --fortesting --single_pass --psuedonames');
+  timedExecOrDie('gulp test --integration ' +
+   '--nobuild --compiled --single_pass --headless');
```
Comment (rsimha) [inline build-system/pr-check/dist-tests.js]: 4 space indent, right?

File: build-system/tasks/presubmit-checks.js [hunk 34]
```diff
@@ -405,7 +405,6 @@ const forbiddenTerms = {
   'isTrustedViewer': {
     message: requiresReviewPrivacy,
     whitelist: [
-      'extensions/amp-bind/0.1/bind-impl.js',
```
Comment (rsimha) [inline build-system/tasks/presubmit-checks.js]: Seems like another stray change that doesn't belong her?

File: .travis.yml [hunk 35]
```diff
@@ -33,21 +27,62 @@ branches:
...
+      script:
+        - node build-system/pr-check/local-tests.js
+    - stage: test
+      name: "Remote (Sauce Labs) Tests"
+      script:
+        - node build-system/pr-check/remote-tests.js
+      after_script:
+        - build-system/sauce_connect/stop_sauce_connect.sh
+  fast_finish: true
+  allowed_failures:
```
Comment (rsimha) [inline .travis.yml]: According to https://docs.travis-ci.com/user/customizing-the-build/#rows-that-are-allowed-to-fail, this is `allow_failures`, not `allowed_failures`

For tracking, add a TODO above this section to eventually remove it.

Finally, test this behavior with a follow up dummy PR by injecting a known failure into a test that will run on Sauce Labs.

File: build-system/tasks/presubmit-checks.js [hunk 36]
```diff
@@ -104,6 +104,14 @@ const forbiddenTerms = {
       'build-system/app.js',
       'build-system/amp4test.js',
       'build-system/check-package-manager.js',
+      'build-system/pr-check/build.js',
```
Comment (rsimha) [inline build-system/tasks/presubmit-checks.js]: I think `build-target.js` is missing from this list. (While you're at it, pluralize the filename to `build-targets.js`.)

File: build-system/tasks/presubmit-checks.js (hunk 27 above)
Comment (estherkim) [reply build-system/tasks/presubmit-checks.js]: Found this! https://github.com/ampproject/amphtml/pull/20822/files

File: build-system/pr-check/build.js [hunk 37]
```diff
@@ -0,0 +1,116 @@
...
+        buildTargets.has('UNIT_TEST') ||
+        buildTargets.has('INTEGRATION_TEST') ||
+        buildTargets.has('BUILD_SYSTEM')) {
+
+      timedExecOrDie('gulp update-packages');
+      timedExecOrDie('gulp build --fortesting');
+      uploadBuildOutput(FILENAME);
+    } else {
+      console.log('Skipping build job because this commit does ' +
+       'not affect the runtime, build system, or test files');
```
Comment (rsimha) [inline build-system/pr-check/build.js]: Before merging, sanity check all test jobs to make sure they are correctly handling the case when the build is skipped.

For example, I see that we're running `gulp test --nobuild` for the `DEV_DASHBOARD` case. Does it need `gulp build` to have been run?

File: build-system/pr-check/remote-tests.js [hunk 38]
```diff
@@ -0,0 +1,77 @@
...
+function main() {
+  const startTime = startTimer(FILENAME, FILENAME);
+  const buildTargets = determineBuildTargets();
+  printChangeSummary(FILENAME);
+  downloadBuildOutput(FILENAME);
+  startSauceConnect(FILENAME);
+
+  if (!isTravisPullRequestBuild()) {
+    timedExecOrDie('gulp test --unit --nobuild --saucelabs_lite');
+    timedExecOrDie('gulp test --integration --nobuild --compiled --saucelabs');
```
Comment (rsimha) [inline build-system/pr-check/remote-tests.js]: This is running integration tests in `--compiled` mode (which means it expects `gulp dist --fortesting` to have been run at some point prior). The implementation of `build.js` seems to be breaking that assumption.

One thing you could do to fix this is to run `gulp dist --fortesting` after `gulp build` in `build.js` in the push build case. It will increase the running time of the `Build` job alone to ~30 minutes. We can revisit this in a follow up PR to optimize the timing of push builds.

Comment (estherkim) [reply build-system/pr-check/remote-tests.js]: What if we ran `gulp test --fortesting` in the remote tests job? 

We're running `gulp dist --fortesting --single_pass --psuedonames` and `gulp dist --fortesting --noextensions` in the dist jobs for push builds, so it wouldn't be breaking a pattern. It would also allow for the test stage to start sooner.

Comment (rsimha) [reply build-system/pr-check/remote-tests.js]: I assume you meant...

> What if we ran `gulp dist --fortesting` in the remote tests job?

Yes, that will certainly solve the problem, since it's just the remote tests that run in compiled mode today. This SGTM.

File: build-system/pr-check/remote-tests.js [hunk 39]
```diff
@@ -44,6 +44,7 @@ function main() {
   startSauceConnect(FILENAME);
 
   if (!isTravisPullRequestBuild()) {
+    timedExecOrDie('gulp dist --fortesting');
     timedExecOrDie('gulp test --unit --nobuild --saucelabs_lite');
```
Comment (rsimha) [inline build-system/pr-check/remote-tests.js]: Move `test --unit` above `dist`, since we can save the time spent running `dist` if the unit tests fail.

There's a new risk here, now that remote tests are allowed to fail: If `gulp dist --fortesting` breaks, we'll never catch it. Not sure how to solve this, but let's punt and consult with @erwinmombay.

File: build-system/pr-check/remote-tests.js (hunk 38 above)
Comment (estherkim) [reply build-system/pr-check/remote-tests.js]: Yes gulp dist! Okay great.

File: build-system/pr-check/remote-tests.js (hunk 39 above)
Comment (estherkim) [reply build-system/pr-check/remote-tests.js]: Sounds good.

**Comment by danielrozenberg** [comment]
Add `Fixes #20589` to the description of this PR

File: .travis.yml (hunk 35 above)
Comment (estherkim) [reply .travis.yml]: This fixed it, thanks. Example here - https://travis-ci.org/ampproject/amphtml/builds/493911037

**Comment by estherkim** [comment]
Agreed - tracked the work here https://github.com/ampproject/amphtml/issues/20887
Okay to do in a future PR?

File: build-system/pr-check/build.js (hunk 37 above)
Comment (estherkim) [reply build-system/pr-check/build.js]: Done! Thanks, made a few adjustments.

**Comment by rsimha** [comment]
New matrix looks good: https://travis-ci.org/ampproject/amphtml/builds/493937270

Merging to make sure `push` builds look good too.

**Comment by rsimha** [comment]
@estherkim Here is the first `push` build that uses the new matrix: https://travis-ci.org/ampproject/amphtml/builds/493957830.
